lib/patchelf/alt_saver.rb

```diff
@@ -622,42 +647,169 @@ def rewrite_sections_library
+        else
+          new_phdrs << new_phdr
+        end
+
+        curr_off += size
+      end
+
+      new_phdrs
+    end
+
+    def sections_at_aligned_offset(offset)
+      @sections.each do |sec|
+        shdr = sec.header
+
+        aligned_offset = Helper.alignup(offset, shdr.sh_addralign)
+        next if shdr.sh_offset != aligned_offset
+
+        yield sec
+      end
+    end
+
+    def shift_sections(shift, start_offset)
+      ehdr.e_shoff = ehdr.e_shoff + shift if ehdr.e_shoff >= start_offset
 
       @sections.each_with_index do |sec, i|
         next if i.zero? # dont touch NULL section
 
         shdr = sec.header
+        next if shdr.sh_offset < start_offset
+
         shdr.sh_offset += shift
       end
+    end
 
-      @segments.each do |seg|
+    def shift_segment_offset(phdr, shift)
+      phdr.p_offset += shift
+      phdr.p_align = page_size if phdr.p_align != 0 && (phdr.p_vaddr - phdr.p_offset) % phdr.p_align != 0
+    end
+
+    def shift_segment_virtual_address(phdr, shift)
+      phdr.p_paddr -= shift if phdr.p_paddr > shift
+      phdr.p_vaddr -= shift if phdr.p_vaddr > shift
+    end
+
+    def shift_segments(shift, start_offset)
+      split_index = -1
+      split_shift = 0
+
+      @segments.each_with_index do |seg, idx|
         phdr = seg.header
-        phdr.p_offset += shift
-        phdr.p_align = page_size if phdr.p_align != 0 && (phdr.p_vaddr - phdr.p_offset) % phdr.p_align != 0
+        p_start = phdr.p_offset
+
+        if p_start <= start_offset && p_start + phdr.p_filesz > start_offset &&
+           phdr.p_type == ELFTools::Constants::PT_LOAD
+          raise PatchError, "split_index(#{split_index}) != -1" if split_index != -1
+
+          split_index = idx
+          split_shift = start_offset - p_start
+
+          phdr.p_offset = start_offset
+          phdr.p_memsz -= split_shift
+          phdr.p_filesz -= split_shift
+          phdr.p_paddr += split_shift
+          phdr.p_vaddr += split_shift
+
+          p_start = start_offset
+        end
+
+        if p_start >= start_offset
+          shift_segment_offset(phdr, shift)
+        else
+          shift_segment_virtual_address(phdr, shift)
+        end
       end
 
+      [split_index, split_shift]
+    end
+
+    def shift_file(extra_pages, start_offset)
+      raise PatchError, "start_offset(#{start_offset}) < ehdr.num_bytes" if start_offset < ehdr.num_bytes
+
+      oldsz = @buffer.size
+      raise PatchError, "oldsz <= start_offset(#{start_offset})" if oldsz <= start_offset
+
+      shift = extra_pages * page_size
+      buf_grow!(oldsz + shift)
+      buf_move! start_offset + shift, start_offset, oldsz - start_offset
+      with_buf_at(start_offset) { |buf| buf.write "\x00" * shift }
+
+      ehdr.e_phoff = ehdr.num_bytes
+
+      shift_sections(shift, start_offset)
+
+      split_index, split_shift = shift_segments(shift, start_offset)
+      raise PatchError, "split_index(#{split_index}) == -1" if split_index == -1
```

Comment:
Maybe move this exception into shift_segments?

Reply:
This will trip `shift_segments` into failing RuboCop's `Metrics/PerceivedComplexity`. Will need a bit of refactoring to do this.

Reply:
let's move it into `shift_segments` and mark `shift_segments` with rubocop:disable Metrics/PerceivedComplexity. I'm surprised that function is the only one complained by Rubocop though.

Reply:
> I'm surprised that function is the only one complained by Rubocop though.

Oh there was a lot more - I just managed to split things up enough to fix it all. Some were quite tricky.